--- git_ori/scripts/writesensor.mako.cpp	2018-03-15 17:12:07.787724175 -0400
+++ git/scripts/writesensor.mako.cpp	2018-03-15 17:14:23.819391186 -0400
@@ -12,6 +12,8 @@
         updateFunc = "set::"
     elif serviceInterface == "xyz.openbmc_project.Inventory.Manager":
         updateFunc = "notify::"
+    elif serviceInterface == "xyz.openbmc_project.Sensor.Temperature":
+        updateFunc = "notify::"
     else:
         assert "Un-supported interface: " + serviceInterface
     endif
@@ -65,6 +67,9 @@
        multiplier = sensor.get("multiplierM", 1)
        offset = sensor.get("offsetB", 0)
        exp = sensor.get("bExp", 0)
+       exponentR = sensor.get("rExp",0)
+       R1 = sensor.get("R1",0)
+       R2 = sensor.get("R2",0)
        valueReadingType = sensor["readingType"]
        updateFunc = interfaceDict[serviceInterface]["updateFunc"]
        updateFunc += sensor["readingType"]
@@ -75,7 +80,7 @@
            sensorInterface = next(iter(interfaces))
 %>
         ${sensorType},"${path}","${sensorInterface}",${readingType},${multiplier},
-        ${offset},${exp},${offset * pow(10,exp)},${updateFunc},{
+        ${offset},${exp},${exponentR},${R1},${R2},${updateFunc},{
     % for interface,properties in interfaces.items():
             {"${interface}",{
             % for dbus_property,property_value in properties.items():
