--- git-orig/arch/arm/include/asm/setup.h	2018-01-31 09:48:11.025610067 +0000
+++ git/arch/arm/include/asm/setup.h	2018-02-05 14:13:50.948687385 +0000
@@ -205,6 +205,24 @@
 	u32 fmemclk;
 };
 
+#define ATAG_ENETADDR	0x41000901
+struct tag_enetaddr {
+	unsigned long enet_count;
+	unsigned char enet0_addr[6];
+	unsigned char enet1_addr[6];
+#ifdef CONFIG_HAS_ETH2	
+	unsigned char enet2_addr[6];
+#endif	
+#ifdef CONFIG_HAS_ETH3
+	unsigned char enet3_addr[6];
+#endif	
+#ifdef CONFIG_VIRTUAL_ETH	
+	unsigned char vnet0_addr[6];
+	unsigned char vnet1_addr[6];
+	unsigned char vnet2_addr[6];
+	unsigned char vnet3_addr[6];
+#endif	
+};
 struct tag {
 	struct tag_header hdr;
 	union {
@@ -227,6 +245,7 @@
 		 * DC21285 specific
 		 */
 		struct tag_memclk	memclk;
+		struct tag_enetaddr	enetaddr;
 	} u;
 };
 
