--- git-orig/arch/arm/lib/interrupts.c	2018-01-31 09:48:11.069610066 +0000
+++ git/arch/arm/lib/interrupts.c	2018-02-05 15:15:39.344580401 +0000
@@ -212,6 +212,7 @@
 	bad_mode ();
 }
 
+
 #ifndef CONFIG_USE_IRQ
 void do_irq (struct pt_regs *pt_regs)
 {
@@ -220,4 +221,10 @@
 	show_regs (pt_regs);
 	bad_mode ();
 }
+#else
+extern void HandleIntr(void);
+void do_irq (struct pt_regs *pt_regs)
+{
+	HandleIntr();
+}
 #endif
